visualization/app/codeCharta/state/store/metricData/edgeMetricData/edgeMetricData.service.ts

```diff
@@ -0,0 +1,111 @@
+import { FilesSelectionSubscriber, FilesService } from "../../files/files.service"
+import { nodeEdgeMetricsMap } from "./edgeMetricData.reducer"
+import { AttributeTypesService, AttributeTypesSubscriber } from "../../fileSettings/attributeTypes/attributeTypes.service"
+
+export interface EdgeMetricDataSubscriber {
+	onEdgeMetricDataChanged(edgeMetricData: EdgeMetricData[])
+}
+
+export class EdgeMetricDataService implements StoreSubscriber, BlacklistSubscriber, FilesSelectionSubscriber, AttributeTypesSubscriber {
+	private static EDGE_METRIC_DATA_CHANGED_EVENT = "edge-metric-data-changed"
+	public static NONE_METRIC = "None"
+
+	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
+		StoreService.subscribe(this.$rootScope, this)
+		BlacklistService.subscribe(this.$rootScope, this)
+		FilesService.subscribe(this.$rootScope, this)
+		AttributeTypesService.subscribe(this.$rootScope, this)
+	}
+
+	public onStoreChanged(actionType: string) {
+		if (isActionOfType(actionType, EdgeMetricDataActions)) {
+			this.notify(this.select())
+		}
+	}
+
+	public onBlacklistChanged(blacklist: BlacklistItem[]) {
+		this.storeService.dispatch(calculateNewEdgeMetricData(this.storeService.getState().files, blacklist))
+	}
+
+	public onFilesSelectionChanged(files: FileState[]) {
+		this.storeService.dispatch(calculateNewEdgeMetricData(files, this.storeService.getState().fileSettings.blacklist))
+	}
+
+	public onAttributeTypesChanged(attributeTypes: AttributeTypes) {
+		this.storeService.dispatch(
+			calculateNewEdgeMetricData(this.storeService.getState().files, this.storeService.getState().fileSettings.blacklist)
+		)
+	}
+
+	public getMetricNames(): string[] {
+		if (!this.storeService.getState().metricData.edgeMetricData) {
+			return []
+		}
+
+		return this.storeService.getState().metricData.edgeMetricData.map(x => x.name)
+	}
+
+	public getAmountOfAffectedBuildings(metricName: string): number {
+		if (!nodeEdgeMetricsMap || !nodeEdgeMetricsMap.has(metricName)) {
+			return 0
+		}
+		return nodeEdgeMetricsMap.get(metricName).size
+	}
+
+	public getNodesWithHighestValue(metricName: string, numberOfNodes: number): string[] {
+		if (!nodeEdgeMetricsMap || !nodeEdgeMetricsMap.has(metricName)) {
+			return []
+		}
+
+		const highestEdgeCountBuildings: string[] = []
+		const edgeMetricMapKeyIterator = nodeEdgeMetricsMap.get(metricName).keys()
+		for (let i = 0; i < numberOfNodes; i++) {
+			highestEdgeCountBuildings.push(edgeMetricMapKeyIterator.next().value)
+		}
+
+		return highestEdgeCountBuildings
+	}
+
+	public getMetricValuesForNode(node: HierarchyNode<CodeMapNode>): Map<string, EdgeMetricCount> {
+		const metricNames = this.getMetricNames().filter(it => !!nodeEdgeMetricsMap.get(it))
+		const nodeEdgeMetrics = new Map()
+
+		metricNames.forEach(metric => {
+			nodeEdgeMetrics.set(metric, nodeEdgeMetricsMap.get(metric).get(node.data.path))
+		})
```

Comment:
```suggestion
		const nodeEdgeMetrics = new Map()
		this.getMetricNames().forEach(metric => {
			const edgeMetrics = nodeEdgeMetricsMap.get(metric)
			if (edgeMetrics === undefined)
				nodeEdgeMetrics.set(metric, edgeMetrics.get(node.data.path))
		})
```

Doing a `filter` and a `forEach` after each other is not good performance wise. Instead, just do the latter and do the filtering in the forEach.
```

Reply:
Absolutely. However, the if-condition should be `if(edgeMetric !== undefined)`.
`!!nodeEdgeMetricsMap.get(it)` is the same as `nodeEdgeMetricsMap.has(it))`

Reply:
> the if-condition should be if(edgeMetric !== undefined)

Indeed!

> !!nodeEdgeMetricsMap.get(it) is the same as nodeEdgeMetricsMap.has(it))

That is actually not correct. The first part checks for in-existent values and values that are falsy. The latter only checks for in-existent values.